### PR TITLE
TRT-1274: get azure disk metrics and create relevant intervals

### PR DESCRIPTION
**Pull request by xueqzhan**
[TRT-1274](https://issues.redhat.com/browse/TRT-1274)

This PR creates a framework for fetching azure metrics and creates intervals based on certain threshold. The goal is to add more diagnosing signals to spyglass to correlate disruption with. The first use case is to correlate etcd slowness with azure disk queue depth.

### CODE REVIEW

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

**Comment by xueqzhan** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-azure-ovn

File: pkg/monitortests/cloud/azure/metrics/monitortest.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+		return nil, nil, err
+	}
+	subscriptionID := config.SubscriptionID
+
+	azureutil.ExportAzureCredentials()
+
+	// create azure metrics client
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		logrus.Errorf("Default azure credential does not exist: %+v", err)
```
Comment (dgoodwin) [inline pkg/monitortests/cloud/azure/metrics/monitortest.go:165]: WithError

**Review by dgoodwin — COMMENTED** [review]
This is extremely cool. Nice job figuring out all the new use of structured intervals as well.

File: pkg/monitortests/cloud/azure/metrics/monitortest.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+			for _, value := range resp.Value {
+				for _, ts := range value.Timeseries {
+					for _, d := range ts.Data {
+						if d.Average != nil && *d.Average > test.avgThreshold {
+							message := fmt.Sprintf("Average value of %f for metric %s is over the threshold of %f", *d.Average, metric, test.avgThreshold)
+							ret = append(ret, monitorapi.NewInterval(monitorapi.SourceCloudMetrics, monitorapi.Error).
+								Locator(monitorapi.NewLocator().CloudNodeMetric(machineName, metric)).
+								Message(monitorapi.NewMessage().Reason(monitorapi.IntervalReason(monitorapi.CloudMetricsExtrenuous)).HumanMessage(message)).
+								Display().
+								Build(d.TimeStamp.Add(-1*time.Minute), *d.TimeStamp),
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/monitortests/cloud/azure/metrics/monitortest.go:120]: Hmm I see, so effectively we probably shouldn't combine  the intervals at all then? That makes sense.

[the remark this answers is not in the document]
Comment (xueqzhan) [reply pkg/monitortests/cloud/azure/metrics/monitortest.go:120]: Yeah, but we might adjust the threshold based on real values to cut down noises.

**Comment by flavianmissi** [comment]
Hi @xueqzhan,
This code isn't working on Azure Stack Hub:
```
[Jira:"Test Framework"] monitor test azure-metrics-collector collection expand_less 	2s
{  failed during collection
GET https://management.azure.com/subscriptions/de7e09c3-b59a-4c7d-9c77-439c11b92879/resourceGroups/ci-op-xgt2820t-195eb/providers/Microsoft.Compute/virtualMachines/ci-op-xgt2820t-195eb-nhgsl-master-0/providers/Microsoft.Insights/metrics
--------------------------------------------------------------------------------
RESPONSE 404: 404 Not Found
ERROR CODE: SubscriptionNotFound
--------------------------------------------------------------------------------
{
  "error": {
    "code": "SubscriptionNotFound",
    "message": "The subscription 'de7e09c3-b59a-4c7d-9c77-439c11b92879' could not be found."
  }
}
--------------------------------------------------------------------------------
}
```
CI job: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7768/pull-ci-openshift-installer-master-e2e-azurestack/1729072030793338880
---
I had the same exact problem on one of my PRs and the solution on my end was to create client options for the client being used:
```
	clientOpts := &arm.ClientOptions{
		ClientOptions: azcore.ClientOptions{
			Cloud: cloud.Configuration{
				ActiveDirectoryAuthorityHost: environment.ActiveDirectoryEndpoint,
				Services: map[cloud.ServiceName]cloud.ServiceConfiguration{
					cloud.ResourceManager: {
						Audience: environment.TokenAudience,
						Endpoint: environment.ResourceManagerEndpoint,
					},
				},
			},
		},
	}
	clientFactory, err := armmonitor.NewClientFactory(subscriptionID, cred, clientOpts)
        // ...
```
One thing to take into consideration however is that the monitor api might not be available on Azure Stack Hub, in which case afaik the error will seem to be permission related. If this is the case my suggestion is to skip this code entirely when running on Azure Stack Hub.
Are you available to provide a fix?
Thanks!

**Comment by xueqzhan** [comment]
> on Azure Stack Hub

@flavianmissi Thank you for pointing this out. I created a card for [TRT-1373](https://issues.redhat.com/browse/TRT-1373) for further investigation/fix on this. For now, we are ignoring all azure interaction errors because the metrics collection is simply there to help debugging and those errors are not fatal for the jobs. The PR for ignoring errors are merging soon: https://github.com/openshift/origin/pull/28420.

**Comment by flavianmissi** [comment]
Understood, thank you @xueqzhan !

**Pull request merged**